src/v/kafka/server/handlers/produce.cc

```diff
@@ -214,55 +236,104 @@ static ss::future<produce_response::partition> produce_topic_partition(
     auto num_records = batch.record_count();
     auto reader = reader_from_lcore_batch(std::move(batch));
     auto start = std::chrono::steady_clock::now();
-    auto f = octx.rctx.partition_manager().invoke_on(
-      *shard,
-      octx.ssg,
-      [reader = std::move(reader),
-       ntp = std::move(ntp),
-       num_records,
-       bid,
-       acks = octx.request.data.acks](cluster::partition_manager& mgr) mutable {
-          auto partition = mgr.get(ntp);
-          if (!partition) {
-              return ss::make_ready_future<produce_response::partition>(
-                produce_response::partition{
-                  .partition_index = ntp.tp.partition,
-                  .error_code = error_code::unknown_topic_or_partition});
-          }
-          if (unlikely(!partition->is_leader())) {
-              return ss::make_ready_future<produce_response::partition>(
-                produce_response::partition{
-                  .partition_index = ntp.tp.partition,
-                  .error_code = error_code::not_leader_for_partition});
-          }
-          return partition_append(
-            ntp.tp.partition,
-            ss::make_lw_shared<replicated_partition>(std::move(partition)),
-            bid,
-            std::move(reader),
-            acks,
-            num_records);
-      });
-    return f.then([&octx, start](produce_response::partition p) {
-        if (p.error_code == error_code::none) {
-            auto dur = std::chrono::steady_clock::now() - start;
-            octx.rctx.connection()->server().update_produce_latency(dur);
-        }
-        return p;
-    });
+
+    auto dispatch = ss::make_foreign<std::unique_ptr<ss::promise<>>>(
+      std::make_unique<ss::promise<>>());
+    auto dispatch_f = dispatch->get_future();
+    auto f
+      = octx.rctx.partition_manager()
+          .invoke_on(
+            *shard,
+            octx.ssg,
+            [reader = std::move(reader),
+             ntp = std::move(ntp),
+             dispatch = std::move(dispatch),
+             num_records,
+             bid,
+             acks = octx.request.data.acks,
+             source_shard = ss::this_shard_id()](
+              cluster::partition_manager& mgr) mutable {
+                auto partition = mgr.get(ntp);
+                if (!partition) {
+                    // submit back to promise source shard
+                    return ss::smp::submit_to(
+                      source_shard,
+                      [dispatch = std::move(dispatch),
+                       pid = ntp.tp.partition]() mutable {
+                          dispatch->set_value();
```

Comment:
it seems like here (and below for not-leader case) you could run the `submit_to` in the background and all it would do is set the value on the dispatch promise. that would probably avoid a full cross cpu round trip? OTOH this is an error case so its not a huge deal. I'm more interested in if there is something that prevents it from running in the background?

Reply:
good idea, I will change this.